### ISSUE TITLE
Removing rows in bigger than memory datasets

### ISSUE THREAD

**Issue by glrs** [linnarsson-lab/loompy#59]
I have a .loom file of 33694 rows (genes) x 632571 columns (cells), so keep in mind that this volume of data cannot fit into the available memory. I want to do some gene filtering and thus removing some of them. I know that for performance reasons, deleting rows directly is not allowed, so I tried to use the `loompy.scan()` for `axis=0`. Here's a minimal example:

```
import loompy
import numpy as np

ds = loompy.connect('my_data.loom', 'r')

cg_exp = ds.map([np.count_nonzero], axis=0)

genes = np.where(cg_exp > 2)[0]
for (ix, selection, view) in ds.scan(items=genes, axis=0):
    loompy.create_append('gene_filtered_ds.loom', view.layers, view.ra, view.ca)
```

This throws:

```
/home/user/anaconda3/envs/tensorflow/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Traceback (most recent call last):
  File "gene_filtering.py", line 10, in <module>
    loompy.create_append('gene_filtered_ds.loom', view.layers, view.ra, view.ca)
  File "/home/user/anaconda3/envs/tensorflow/lib/python3.6/site-packages/loompy/loompy.py", line 807, in create_append
    ds.add_columns(layers, col_attrs, fill_values=fill_values)
  File "/home/user/anaconda3/envs/tensorflow/lib/python3.6/site-packages/loompy/loompy.py", line 281, in add_columns
    raise ValueError(f"Layer {layer} has {matrix.shape[0]} rows but file has {self.shape[0]}")
ValueError: Layer  has 460 rows but file has 415
```

I can see two issues in my code:

1. the `loompy.create_append()` tries to append columns to the file, which is of course the wrong way to go,
2. it tries to match the rows, but the genes in each chunk are different (both as genes per se and as number of genes passed the QC).

I played around with the `loompy.scan()` since from the documentation it seems able to return batches of either rows or columns, however I couldn't find any good way to write the data on a file iteratively. Is there any solution to this?

**Comment by slinnarsson** [comment]
Hi

First, a small issue: `ds.map()` always returns a tuple, so you need to unpack it:

```python
    cg_exp = ds.map([np.count_nonzero], axis=0)[0]
```

The reason for this is that you pass in a tuple of functions, and the returned result is a tuple of the result of those functions.

Second, your example is close, but the trick in this case is to scan horizontally, with axis=1, so that scan will return chunks of columns, and then to manually slice the view to get the rows you want:

```python
import loompy
import numpy as np

f = 'my_data.loom'
with loompy.connect(f, 'r') as ds:
    cg_exp = ds.map([np.count_nonzero], axis=0)[0]
    genes = np.where(cg_exp > 2)[0]
    for (ix, selection, view) in ds.scan(axis=1):
        loompy.create_append('gene_filtered_ds.loom', view.layers[genes, :], view.ra[genes], view.ca)
```

Also note that it's always recommended to use a `with` statement, so that the file is automatically closed for you. This prevents many HDF5 errors.

Finally, I always tend to mark genes rather then remove them. So you might do this instead:

```python
import loompy
import numpy as np

f = 'my_data.loom'
with loompy.connect(f, 'r+') as ds:
    cg_exp = ds.map([np.count_nonzero], axis=0)[0]
    ds.ra.Selected = cg_exp > 2
```
(Note the mode must be `r+` since you will be writing to the file). This has the advantage that you will always be able to plot all genes, even if they were not selected. You just need to remember downstream to always filter the genes by the `Selected` attribute. And be careful when you read the attribute - HDF5 doesn't support booleans so it will be returned as an array of 1s and 0s. You need to do `ds.ra.Selected == 1` to get a boolean array back.

**Comment by glrs** [comment]
Thanks, it worked I will close this "issue" since it wasn't really one. However, in my attempts to make this work I tried to transpose the whole file, cut the columns (genes), then transpose it back again. Doing that, I realized that saving the transposed file consumes way less space than the original one. Also, trying to transpose it back, the volume inflates multiple times the original file (even though I applied the suggestion of issue #49 ). Does this still has to do with the distribution of zeros and the zero chunks?
